test/library/packages/UnitTest/Launcher.chpl

```diff
@@ -21,86 +21,153 @@ module Launcher {
+      }
+      else {
+        writeln("chpl not found.");
+      } 
     }
     
-    var result =  new TestResult();
-
-    for tests in files {
-      try {
-        testFile(tests, result);
+  }
+  
+  pragma "no doc"
+  /*Docs: Todo*/
+  proc testFile(file, ref result) throws {
+    var fileName = basename(file);
+    if fileName.endsWith(".chpl") {
+      var line: string;
+      var compErr = false;
+      var tempName = fileName.split(".chpl");
+      var executable = tempName[1];
+      var executableReal = executable + "_real";
+      if isFile(executable) {
+        FileSystem.remove(executable);
       }
-      catch e {
-        writeln("Caught an Exception in Running Test File: ", tests);
-        writeln(e);
+      if isFile(executableReal) {
+        FileSystem.remove(executableReal);
       }
-    }
-
-    for dir in dirs {
-      try {
-        testDirectory(dir, result);
+      if setComm != "" then comm = setComm;
+      var sub = spawn(["chpl", file, "-o", executable, "-M.", 
+                      "--comm", comm], stderr = PIPE); //Compiling the file
+      var compError: string;
+      if sub.stderr.readline(line) {
+        compError = line;
+        while sub.stderr.readline(line) do compError += line;
+        compErr = true;
       }
-      catch e {
-        writeln("Caught an Exception in Running Test Directory: ", dir);
-        writeln(e);
+      sub.wait();
+      if !compErr {
+        var testNames: [1..0] string;
+        var dictDomain: domain(int);
+        var dict: [dictDomain] int;
+        runAndLog(executable, fileName, result, 0, numLocales,
+                  testNames, dictDomain, dict);
+        if !keepExec {
+          FileSystem.remove(executable);
+          if isFile(executableReal) {
+            FileSystem.remove(executableReal);
+          }
+        }
+      }
+      else {
+        writeln("Compilation Error in ", fileName);
+        writeln("Possible Reasons can be passing a non-test ",
+                "function to UnitTest.runTest()");
+        writeln(compError);
       }
     }
-      
-    result.printErrors();
-    writeln(result.separator2);
-    result.printResult();
   }
 
   pragma "no doc"
   /*Docs: Todo*/
-  proc runAndLog(executable,fileName, ref result,skipId = 0,
-                  reqNumLocales:int = numLocales, ref testNames,
-                  ref dictDomain,ref dict) throws {
+  proc testDirectory(dir, ref result) throws {
+    for file in findfiles(startdir = dir, recursive = subdir) {
+      testFile(file, result);
+    }
+  }
+
+  pragma "no doc"
+  /*Docs: Todo*/
+  proc runAndLog(executable, fileName, ref result, skipId = 0,
+                  reqNumLocales: int = numLocales, ref testNames,
+                  ref dictDomain, ref dict) throws 
+  {
     var separator1 = result.separator1,
         separator2 = result.separator2;
     var testName: string,
         flavour: string,
         line: string,
-        tempString: string;
+        testExecMsg: string;
```

Comment:
This variable gets reset for each iteration of the while loop.  Maybe it should live in the while loop itself?

Reply:
Hey @lydia-duncan these variables are reset only in two conditions out of 4 that's why I don't think we should declare them inside.

Reply:
But it is only used in those ones that reset it

Reply:
Is there a situation where you would want it to keep the previous value?

Reply:
Yes, Ma'am. This variable is storing the each line while we read the lines through stdout. In two cases it is reset and in one case it is getting updated.
```chapel
else if sep1Found then testExecMsg += line;
```

Reply:
Oh, on line 184!  That if/else confused me, sorry about that!